### PR TITLE
Refactoring: Resource Provider Registration / Auth prep.

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR refactors the Resource Provider registration out into it's own package; it also moves some of the Authentication related stuff out into it's own package since it's going to be refactored further in the near future.

As a part of refactoring the Resource Provider registration out, I've also inverted the dependency such that we're able to reuse the RP Registration once it's pulled out it's own library - which will become part of the greater shared authentication library (across AzureRM/Public, Azure Stack and the AzureRM Backend in Terraform Core)

In the near future the intention is to have a shared authentication package with feature toggles for AzureCLIParsingAuth, ClientSecretAuth, CloudShellAuth etc - so that we're able to decide what's supported in each location as necessary (for example, CloudShell isn't available in Azure Stack); so whilst some of this PR may seem superfluous, there's some prep work for the upcoming refactor

### CODE REVIEW

File: azurerm/helpers/authentication/config.go [hunk 1]
```diff
@@ -9,10 +9,12 @@ import (
 	"github.com/Azure/go-autorest/autorest/azure/cli"
 )
 
+// TODO: separate objects for Input/Output
```
Comment (bflad) [inline azurerm/helpers/authentication/config.go]: Should there be a GitHub issue created for these TODOs?

File: azurerm/helpers/resourceproviders/registration.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2017-05-10/resources"
+)
+
+func DetermineResourceProvidersRequiringRegistration(availableResourceProviders []resources.Provider, requiredResourceProviders map[string]struct{}) map[string]struct{} {
```
Comment (bflad) [inline azurerm/helpers/resourceproviders/registration.go]: Nit: Exported functions should be documented

File: azurerm/helpers/azure/sender.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"log"
+	"net/http"
+	"net/http/httputil"
+
+	"github.com/Azure/go-autorest/autorest"
+)
+
+func BuildSender() autorest.Sender {
```
Comment (bflad) [inline azurerm/helpers/azure/sender.go]: Nit: Exported functions should be documented

File: azurerm/helpers/authentication/validation.go [hunk 4]
```diff
@@ -2,11 +2,54 @@ package authentication
...
 import (
 	"fmt"
+	"log"
 
+	"github.com/Azure/go-autorest/autorest/adal"
 	"github.com/hashicorp/go-multierror"
 )
 
-func (c *Config) ValidateBearerAuth() error {
+func (c *Config) Validate() error {
```
Comment (bflad) [inline azurerm/helpers/authentication/validation.go]: Nit: Exported functions should be documented

File: azurerm/helpers/authentication/authorization_token.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package authentication
+
+import (
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+)
+
+func GetAuthorizationToken(c *Config, oauthConfig *adal.OAuthConfig, endpoint string) (*autorest.BearerAuthorizer, error) {
```
Comment (bflad) [inline azurerm/helpers/authentication/authorization_token.go]: Nit: Exported functions should be documented

File: azurerm/helpers/authentication/environment.go [hunk 6]
```diff
@@ -1,6 +1,27 @@
...
 
-import "strings"
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+)
+
+func DetermineEnvironment(name string) (*azure.Environment, error) {
```
Comment (bflad) [inline azurerm/helpers/authentication/environment.go]: Nit: Exported functions should be documented

**Review by bflad — APPROVED** [review]
LGTM once passing testing is shown and potentially documenting/unexporting some of these functions. 🚀

File: azurerm/helpers/authentication/config.go (hunk 1 above)
Comment (tombuildsstuff) [reply azurerm/helpers/authentication/config.go]: they'll be coming in the next PR - so it should be fine

File: azurerm/helpers/resourceproviders/registration.go (hunk 2 above)
Comment (tombuildsstuff) [reply azurerm/helpers/resourceproviders/registration.go]: done 👍

File: azurerm/helpers/azure/sender.go (hunk 3 above)
Comment (tombuildsstuff) [reply azurerm/helpers/azure/sender.go]: done 👍

File: azurerm/helpers/authentication/validation.go (hunk 4 above)
Comment (tombuildsstuff) [reply azurerm/helpers/authentication/validation.go]: done 👍

File: azurerm/helpers/authentication/authorization_token.go (hunk 5 above)
Comment (tombuildsstuff) [reply azurerm/helpers/authentication/authorization_token.go]: done 👍

File: azurerm/helpers/authentication/environment.go (hunk 6 above)
Comment (tombuildsstuff) [reply azurerm/helpers/authentication/environment.go]: done 👍

**Comment by tombuildsstuff** [comment]
Resource Provider Registration test passes:

Tests pass:

```
$ acctests azurerm TestAccAzureRMEnsureRequiredResourceProvidersAreRegistered
=== RUN   TestAccAzureRMEnsureRequiredResourceProvidersAreRegistered
--- PASS: TestAccAzureRMEnsureRequiredResourceProvidersAreRegistered (5.90s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	7.278s
```

**Pull request merged**